proposals/csharp-7.0/digit-separators.md

```diff
@@ -1,21 +1,181 @@
 # Digit separators
 
-Being able to group digits in large numeric literals would have great readability impact and no significant downside. 
+This proposal specifies the changes required to the [C# 6.0 (draft) Language specification](../../spec/introduction.md) to support *literal digit separators*.
 
-Adding binary literals (#215) would increase the likelihood of numeric literals being long, so the two features enhance each other. 
+## Changes to [Lexical structure](../../spec/lexical-structure.md)
 
-We would follow Java and others, and use an underscore `_` as a digit separator. It would be able to occur everywhere in a numeric literal (except as the first and last character), since different groupings may make sense in different scenarios and especially for different numeric bases:
+### Literals
 
+#### Integer literals
+
+> The grammar for [integer literals](../../spec/lexical-structure.md#Integer-literals) is modified to allow one or more `_` separators anywhere between the first and last digits.
+
+Integer literals are used to write values of types `int`, `uint`, `long`, and `ulong`. Integer literals have three possible forms: decimal, hexadecimal, and binary.
+
+```antlr
+integer_literal
+    : decimal_integer_literal
+    | hexadecimal_integer_literal
+    | binary_integer_literal
+   ;
+
+integer_type_suffix
+    : 'U' | 'u' | 'L' | 'l' | 'UL' | 'Ul' | 'uL' | 'ul' | 'LU' | 'Lu' | 'lU' | 'lu'
+    ;
+
+decimal_integer_literal
+    : decimal_digits integer_type_suffix?
+    ;
+
+decimal_digits:
+    : decimal_digit
+    | decimal_digit decimal_digits_and_underscores? decimal_digit
+    ;
+
+decimal_digit
+    : '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9'
+    ;
+
+decimal_digits_and_underscores:
+    : decimal_digit_or_underscore+
+    ;
+
+decimal_digit_or_underscore:
+    : decimal_digit
+    | '_'
+    ;
+
+hexadecimal_integer_literal
+    : '0x' hex_digits integer_type_suffix?
+    | '0X' hex_digits integer_type_suffix?
+    ;
+
+hex_digits:
+    : hex_digit
+    | hex_digit hex_digits_and_underscores? hex_digit
+    ;
+
+hex_digit
+    : '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9'
+    | 'A' | 'B' | 'C' | 'D' | 'E' | 'F' | 'a' | 'b' | 'c' | 'd' | 'e' | 'f'
+    ;
+
+hex_digits_and_underscores:
+    : hex_digit_or_underscore+
+    ;
+
+hex_digit_or_underscore:
+    : hex_digit
+    | '_'
+    ;
+
+binary_integer_literal:
+    : `0b` binary_digits integer_type_suffix?
+    | `0B` binary_digits integer_type_suffix?
+    ;
```

Comment:
💭 Was this modified to allow `_` to appear between the prefix `0b` and the first binary digit?

Reply:
Not yet; this initial PR is for the 7.0 features, whereas leading underscores are in 7.2. That said, I have privately written up this extra feature, and it will become a PR "real soon now."